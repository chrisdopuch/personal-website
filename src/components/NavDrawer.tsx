import Divider from '@material-ui/core/Divider';
import Drawer from '@material-ui/core/Drawer';
import Hidden from '@material-ui/core/Hidden';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import { createStyles, MuiThemeProvider, Theme, withStyles, WithStyles } from '@material-ui/core/styles';
import { SvgIconProps } from '@material-ui/core/SvgIcon';
import React, { ComponentType, SFC } from 'react';
import { useStore } from 'react-hookstore';
import { Link } from 'react-router-dom';
import { Follow } from 'react-twitter-widgets';
import { appTheme } from './App';

const drawerWidth = 240;

const stylesDeclarations = (theme: Theme) => {
  return createStyles({
    drawer: {
      [theme.breakpoints.up('md')]: {
        flexShrink: 0,
        width: drawerWidth,
      },
    },
    drawerPaper: {
      width: drawerWidth,
    },
    social: {
      margin: '20px 10px',
      textAlign: 'left',
    },
    toolbar: theme.mixins.toolbar,
    twitter: {
      marginBottom: '10px',
    },
  });
};

interface INavDrawerProps extends WithStyles<typeof stylesDeclarations> {
  items: Array<{
    to: string;
    label: string;
    Icon: ComponentType<SvgIconProps>;
  }>;
  theme: Theme;
}

export const NavDrawer: SFC<INavDrawerProps> = (props) => {
  const { classes, theme } = props;
  const [isMobileDrawerOpen, setisMobileDrawerOpen] = useStore();

  return (
    <MuiThemeProvider theme={theme}>
      <nav className={classes.drawer}>
        {/* The implementation can be swap with js to avoid SEO duplication of links. */}
        <Hidden mdUp={true} implementation="css">
          <Drawer
            variant="temporary"
            anchor={theme.direction === 'rtl' ? 'right' : 'left'}
            open={isMobileDrawerOpen}
            onClose={() => setisMobileDrawerOpen(false)} // tslint:disable-line
            classes={{
              paper: classes.drawerPaper,
            }}
            ModalProps={{
              keepMounted: true, // Better open performance on mobile.
            }}
          >
            <DrawerContent {...props} />
          </Drawer>
        </Hidden>
        <Hidden smDown={true} implementation="css">
          <Drawer
            classes={{
              paper: classes.drawerPaper,
            }}
            variant="permanent"
            open={true}
          >
            <DrawerContent {...props} />
          </Drawer>
        </Hidden>
      </nav>
    </MuiThemeProvider>
  );
};

NavDrawer.defaultProps = {
  theme: appTheme,
};

const DrawerContent: SFC<INavDrawerProps> = (props) => {
  const { classes, items } = props;
  const setisMobileDrawerOpen = useStore()[1];

  return (
    <div>
      <div className={classes.toolbar} />
      <Divider />
      <List>
        {items.map(({ label, to, Icon }) => {
          const linkProps = { to } as any;
          return (
            <ListItem
              onClick={() => setisMobileDrawerOpen(false)} // tslint:disable-line
              button={true}
              key={label}
              component={Link}
              {...linkProps}
            >
              <ListItemIcon>
                <Icon />
              </ListItemIcon>
              <ListItemText primary={label} />
            </ListItem>
          );
        })}
      </List>
      <Divider />
      <div className={classes.social}>
        <div className={classes.twitter}>
          <Follow username="chrisdopuch" />
        </div>
        <a className="github-button" href="https://github.com/chrisdopuch" aria-label="Follow @chrisdopuch on GitHub">
          Follow @chrisdopuch
        </a>
      </div>
    </div>
  );
};

export default withStyles(stylesDeclarations, { withTheme: true })(NavDrawer);
