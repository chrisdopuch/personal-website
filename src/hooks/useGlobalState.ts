import { useContext } from 'react';
import { stateCtx } from '../components/contexts/state';
import { State } from '../store';

const useGlobalState = <K extends keyof State>(key: K) => {
  const state = useContext(stateCtx);
  return state[key];
};

export default useGlobalState;
