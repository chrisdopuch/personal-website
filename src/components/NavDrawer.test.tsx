import { Drawer, Theme } from '@material-ui/core';
import { shallow } from 'enzyme';
import React from 'react';
import useGlobalState from '../hooks/useGlobalState';
import { NavDrawer } from './NavDrawer';

const mockDispatch = jest.fn();
jest.mock('../hooks/useDispatch', () => {
  return jest.fn(() => mockDispatch);
});

jest.mock('../hooks/useGlobalState', () => {
  return jest.fn(() => false);
});

function getDefaultProps() {
  return {
    classes: {
      drawer: 'drawer',
      drawerPaper: 'drawerPaper',
    },
    items: [],
    theme: {} as Theme,
  };
}

describe('NavDrawer', () => {
  it('renders with default props', () => {
    const wrapper = shallow(<NavDrawer {...getDefaultProps()} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders with `isNavDrawerOpen` true', () => {
    (useGlobalState as jest.Mock).mockImplementationOnce(() => true);
    const wrapper = shallow(<NavDrawer {...getDefaultProps()} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders with items', () => {
    const MockHomeIcon = jest.fn();
    const MockHelpIcon = jest.fn();
    const MockHomePage = jest.fn();
    const MockAboutPage = jest.fn();
    const props = {
      ...getDefaultProps(),
      items: [
        { to: '/', label: 'Home', Icon: MockHomeIcon, isRouteExact: true, page: MockHomePage },
        { to: '/about', label: 'About me', Icon: MockHelpIcon, isRouteExact: false, page: MockAboutPage },
      ],
    };
    const wrapper = shallow(<NavDrawer {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("dispatches `setNavDrawerOpen` action on Drawer's `onClose`", () => {
    const wrapper = shallow(<NavDrawer {...getDefaultProps()} />);

    wrapper
      .find(Drawer)
      .first()
      .props()
      .onClose();

    expect(mockDispatch).toBeCalledWith({ type: 'setNavDrawerOpen', payload: false });
  });
});
