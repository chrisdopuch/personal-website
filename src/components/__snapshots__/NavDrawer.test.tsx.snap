// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavDrawer snapshots renders default props 1`] = `
<nav
  className="drawer"
>
  <Hidden
    implementation="css"
    lgDown={false}
    lgUp={false}
    mdDown={false}
    mdUp={true}
    smDown={false}
    smUp={false}
    xlDown={false}
    xlUp={false}
    xsDown={false}
    xsUp={false}
  >
    <WithStyles(Drawer)
      ModalProps={
        Object {
          "keepMounted": true,
        }
      }
      anchor="left"
      classes={
        Object {
          "paper": "drawerPaper",
        }
      }
      onClose={[Function]}
      variant="temporary"
    >
      <WithStyles(DrawerContent)
        items={Array []}
      />
    </WithStyles(Drawer)>
  </Hidden>
  <Hidden
    implementation="css"
    lgDown={false}
    lgUp={false}
    mdDown={false}
    mdUp={false}
    smDown={true}
    smUp={false}
    xlDown={false}
    xlUp={false}
    xsDown={false}
    xsUp={false}
  >
    <WithStyles(Drawer)
      classes={
        Object {
          "paper": "drawerPaper",
        }
      }
      open={true}
      variant="permanent"
    >
      <WithStyles(DrawerContent)
        items={Array []}
      />
    </WithStyles(Drawer)>
  </Hidden>
</nav>
`;

exports[`NavDrawer snapshots renders with items 1`] = `
<nav
  className="drawer"
>
  <Hidden
    implementation="css"
    lgDown={false}
    lgUp={false}
    mdDown={false}
    mdUp={true}
    smDown={false}
    smUp={false}
    xlDown={false}
    xlUp={false}
    xsDown={false}
    xsUp={false}
  >
    <WithStyles(Drawer)
      ModalProps={
        Object {
          "keepMounted": true,
        }
      }
      anchor="left"
      classes={
        Object {
          "paper": "drawerPaper",
        }
      }
      onClose={[Function]}
      variant="temporary"
    >
      <WithStyles(DrawerContent)
        items={
          Array [
            Object {
              "Icon": [MockFunction],
              "isRouteExact": true,
              "label": "Home",
              "page": [MockFunction],
              "to": "/",
            },
            Object {
              "Icon": [MockFunction],
              "isRouteExact": false,
              "label": "About me",
              "page": [MockFunction],
              "to": "/about",
            },
          ]
        }
      />
    </WithStyles(Drawer)>
  </Hidden>
  <Hidden
    implementation="css"
    lgDown={false}
    lgUp={false}
    mdDown={false}
    mdUp={false}
    smDown={true}
    smUp={false}
    xlDown={false}
    xlUp={false}
    xsDown={false}
    xsUp={false}
  >
    <WithStyles(Drawer)
      classes={
        Object {
          "paper": "drawerPaper",
        }
      }
      open={true}
      variant="permanent"
    >
      <WithStyles(DrawerContent)
        items={
          Array [
            Object {
              "Icon": [MockFunction],
              "isRouteExact": true,
              "label": "Home",
              "page": [MockFunction],
              "to": "/",
            },
            Object {
              "Icon": [MockFunction],
              "isRouteExact": false,
              "label": "About me",
              "page": [MockFunction],
              "to": "/about",
            },
          ]
        }
      />
    </WithStyles(Drawer)>
  </Hidden>
</nav>
`;
